File: 30.最长递增子序列/30.最长递增子序列.py
"""
动态规划
https://www.bilibili.com/video/BV1XS4y1G7rs/?spm_id_from=autoNext&vd_source=751d83a470a7b7086d96d8b5dd5762d9
"""
class Solution:
    def lengthOfLIS(self, heights: List[int]) -> int:
        nums = len(heights)
        left = [0 for i in range(nums)]

        for i in range(nums):
            left[i] = 1
            for ii in range(i):
                if heights[i] > heights[ii]:
                    left[i] = max(left[i], left[ii] + 1)

        return max(left)

File: 30.最长递增子序列/30.最长递增子序列.cpp
#include <algorithm>
class Solution {
public:
    int lengthOfLIS(vector<int>& heights) {
        int nums=heights.size();
        int height;
        vector<int>left(nums);

        for(int i=0;i<nums;i++)
        {
            left[i] = 1;
            for(int ii=0;ii<i;ii++)
            {
                if(heights[i]>heights[ii])
                {
                    left[i]= max(left[i], left[ii]+1);
                }
            }
        }
        int max_value=0;
        for(auto it=left.begin();it!=left.end();it++){
            if(*it>max_value)
                max_value = *it;
        }
        return max_value;
    }
};